yoyodyne/models/pointer_generator.py

```diff
@@ -444,3 +447,238 @@ def _reshape_hiddens(
+            # we have decoded at least the the same number of steps as the
+            # target length.
+            if finished.all():
+                if targets is None or len(outputs) >= targets.size(-1):
+                    break
+        # -> B x seq_len x target_vocab_size.
+        return torch.stack(outputs).transpose(0, 1)
+    
+    def forward(
+        self,
+        batch: data.PaddedBatch,
+    ) -> torch.Tensor:
+        """Runs the encoder-decoder.
+
+        Args:
+            batch (data.PaddedBatch).
+
+        Returns:
+            torch.Tensor.
+        """
+        source_encoded = self.source_encoder(batch.source).output
+        if self.training and self.teacher_forcing:
+            assert (
+                batch.target.padded is not None
+            ), "Teacher forcing requested but no target provided"
+            # Initializes the start symbol for decoding.
+            starts = (
+                torch.tensor(
+                    [self.start_idx], device=self.device, dtype=torch.long
+                )
+                .repeat(batch.target.padded.size(0))
+                .unsqueeze(1)
+            )
+            target_padded = torch.cat((starts, batch.target.padded), dim=1)
+            target_mask = torch.cat(
+                (starts == self.pad_idx, batch.target.mask), dim=1
+            )
+            if self.has_features_encoder:
+                features_encoder_output = self.features_encoder(batch.features)
+                features_encoded = features_encoder_output.output
+                raise NotImplementedError
+            else:
+                if self.beam_width is not None and self.beam_width > 1:
+                    # predictions = self.beam_decode(
+                    # batch_size, x_mask, encoder_out, beam_width=self.beam_width
+                    # )
+                    raise NotImplementedError
+                else:
+                    scores = self.decode_step(
+                        source_encoded,
+                        batch.source.mask,
+                        batch.source.padded,
+                        target_padded,
+                        target_mask,
+                    )
+                    scores = scores[:, :-1, :]  # Ignore EOS.
+                    # Puts scores in log space.
+                    output = torch.log(scores)
+        else:
+            if self.has_features_encoder:
+                features_encoder_output = self.features_encoder(batch.features)
+                features_encoded = features_encoder_output.output
+                raise NotImplementedError
+            else:
+                # -> B x seq_len x output_size.
+                output = self._decode_greedy(
+                    source_encoded,
+                    batch.source.mask,
+                    batch.source.padded,
+                    batch.target.padded if batch.target else None,
+                )
+        return output
+    
+    @property
+    def name(self) -> str:
```

Comment:
This can be two words, so "pointer-generator transformer" seems fine. Also, you should change the pointer generator LSTM's `name` property to return `pointer-generator LSTM` then I think.